File: rust/src/nodes.rs
pub trait Node {
    fn get_string(&self)->String;
}

pub struct Header {
    pub size: i32,
    pub children: Vec<Box<dyn Node>>
}

pub struct Paragraph {
    pub children: Vec<Box<dyn Node>>
}

pub struct CodeBlock {
    pub text: String
}

pub struct Image {
    pub text: String,
    pub url: String
}

pub struct Text {
    pub text: String
}

pub struct Italic {
    pub children: Vec<Box<dyn Node>>
}

pub struct Bold {
    pub children: Vec<Box<dyn Node>>
}

pub struct Code {
    pub text: String
}

pub struct Link {
    pub text: String,
    pub url: String
}

fn get_child_html_vec(children:&Vec<Box<dyn Node>>)->Vec<String> {
    children.iter().map(|child| child.get_string()).collect()
}

impl Node for Header {
    fn get_string(&self)->String {
        format!("<h{}>{}</h{}>\n", self.size, get_child_html_vec(&self.children).join(""), self.size)
    }
}

impl Node for Paragraph {
    fn get_string(&self)->String {
        format!("<p>{}</p>\n\n", get_child_html_vec(&self.children).join(""))
    }
}

impl Node for CodeBlock {
    fn get_string(&self)->String {
        format!("<pre><code>{}</code></pre>\n\n", self.text)
    }
}

impl Node for Image {
    fn get_string(&self)->String {
        format!("<img src=\"{}\" alt=\"{}\" />\n\n", self.url, self.text)
    }
}

impl Node for Text {
    fn get_string(&self)->String {
        self.text.replace("\r", "")
    }
}

impl Node for Italic {
    fn get_string(&self)->String {
        format!("<em>{}</em>", get_child_html_vec(&self.children).join(""))
    }
}

impl Node for Bold {
    fn get_string(&self)->String {
        format!("<strong>{}</strong>", get_child_html_vec(&self.children).join(""))
    }
}

impl Node for Code {
    fn get_string(&self)->String {
        format!("<code>{}</code>", self.text)
    }
}

impl Node for Link {
    fn get_string(&self)->String {
        format!("<a href=\"{}\">{}</a>", self.url, self.text)
    }
}

File: zig/src/nodes.zig
const std = @import("std");

pub const Node = union(enum) {
    header: struct { size: usize, children: std.ArrayList(Node) },

    paragraph: struct { children: std.ArrayList(Node) },

    codeblock: struct { text: []const u8 },

    image: struct { text: []const u8, url: []const u8 },

    text: struct { text: []const u8 },

    italic: struct { children: std.ArrayList(Node) },

    bold: struct { children: std.ArrayList(Node) },

    code: struct { text: []const u8 },

    link: struct { text: []const u8, url: []const u8 },

    fn printCode(text: []const u8, w: std.fs.File.Writer) !void {
        for (text) |c| {
            if (c != '\r') {
                try w.print("{c}", .{c});
            }
        }
    }

    pub fn getHTML(node: Node, w: std.fs.File.Writer) !void {
        switch (node) {
            .header => {
                try w.print("<h{}>", .{node.header.size});
                for (node.header.children.items) |child| {
                    try getHTML(child, w);
                }
                try w.print("</h{}>\n", .{node.header.size});
            },

            .paragraph => {
                try w.print("{s}", .{"<p>"});
                for (node.paragraph.children.items) |child| {
                    try getHTML(child, w);
                }
                try w.print("{s}", .{"</p>\n\n"});
            },

            .codeblock => {
                try w.print("{s}", .{"<pre><code>"});
                try printCode(node.codeblock.text, w);
                try w.print("{s}", .{"</code></pre>\n\n"});
            },

            .image => try w.print("<img src=\"{s}\" alt=\"{s}\" />\n\n", .{ node.image.url, node.image.text }),

            .text => try w.print("{s}", .{node.text.text}),

            .italic => {
                try w.print("{s}", .{"<em>"});
                for (node.italic.children.items) |child| {
                    try getHTML(child, w);
                }
                try w.print("{s}", .{"</em>"});
            },

            .bold => {
                try w.print("{s}", .{"<strong>"});
                for (node.bold.children.items) |child| {
                    try getHTML(child, w);
                }
                try w.print("{s}", .{"</strong>"});
            },

            .code => {
                try w.print("{s}", .{"<code>"});
                try printCode(node.code.text, w);
                try w.print("{s}", .{"</code>"});
            },

            .link => try w.print("<a href=\"{s}\">{s}</a>", .{ node.link.url, node.link.text }),
        }
    }
};
